### PR TITLE
Refactor scorpio interfaces and classes

### PR DESCRIPTION
**Pull request by bartgol**

### CODE REVIEW

**Comment by bartgol** [comment]
I think this is a good point to stop, and integrate these mods, before moving forward. @AaronDonahue @jeff-cohere , here's a summary of the mods:

- Fix a bug related to how FieldHeader and FieldTracking were tracking parent/children (coming from subfields). The bug was subtle: basically, [`shared_from_this`](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this/shared_from_this) cannot be called from the ctor, since the shared_ptr control block is not yet built, so the returned shared ptr is always null. While fixing this (basically, do the parent-children link _after_ the ctor returns), I thought I would abstract this parent-child link setup/cleanup with a common class, since the logic was the same in both header and tracking classes.
Note: FieldTracking needs the parent/children link _only_ to be able to update the children fields time stamps whenever the parent time stamp is updated. If we think this feature is not useful, we can simplify the FieldTracking class.
- all scorpio I/O classes are now on a single grid. We currently don't need 2+ grids but will soon. The idea will be to have 1 output manager per grid. We will always have very few grids (2 or maybe 3), so this is probably ok. 
- scorpio input: reworked a bit its impl. Can work with a FieldManager (grabs everything from there) or with a vector of 1d views and one of field layouts.
- scorpio output: more rework than for input. Unlike the input class, it only works with FieldManager (we might want to extend to save raw view+layout, perhaps for geometry data). Removed the m_status map, since the only counter really useful is the number of steps since the last output (and only if avg!=Instant). I started to introduce the word "checkpoint" or "output checkpoint" for output history restart files, but I don't think I made it super consistent throughout. Let me know what you think of the wording, and how it can be improved.

I already have a few things in mind for what's next with scorpio (e.g., output (and restart!) on dyn grid, or output non-field quantities like lat/lon). We can discuss those at the telecon.

File: components/scream/src/control/atmosphere_driver.cpp
Comment (AaronDonahue) [inline components/scream/src/control/atmosphere_driver.cpp:226]: This comment reads as though something is missing.

File: components/scream/src/control/atmosphere_driver.cpp
Comment (AaronDonahue) [inline components/scream/src/control/atmosphere_driver.cpp:232]: nitpicky: grid->grids

File: components/scream/src/control/atmosphere_driver.cpp
Comment (jeff-cohere) [inline components/scream/src/control/atmosphere_driver.cpp:323]: Is `skip_init_{lat/lon}` a legacy name? It's less confusing to use terms that indicate things should be done rather than skipped. Skipping things implies assumptions that everyone knows. But if this name is historical, I guess we're stuck with it.

File: components/scream/src/control/atmosphere_driver.hpp
Comment (jeff-cohere) [inline components/scream/src/control/atmosphere_driver.hpp:89]: Can we add some comments above this method that describe the parameters, since we're adding a new one? One can guess at what they mean, but we spend so much time guessing at things this way already. :-)

File: components/scream/src/share/io/scorpio_output.hpp
Comment (AaronDonahue) [inline components/scream/src/share/io/scorpio_output.hpp:162]: I like these new names, they make more sense.

File: components/scream/src/share/io/scorpio_output.hpp
Comment (AaronDonahue) [inline components/scream/src/share/io/scorpio_output.hpp:165]: We should add a comment to this function that explains that it controls updating variables based on the output type and is where any new output types would go.  For example if someone wanted to add standard deviation as an option, or something like that.

File: components/scream/src/share/util/scream_family_tracking.hpp
Comment (jeff-cohere) [inline components/scream/src/share/util/scream_family_tracking.hpp:4]: You've got a double slash here.

File: components/scream/src/share/io/scream_output_manager.cpp
Comment (AaronDonahue) [inline components/scream/src/share/io/scream_output_manager.cpp:94]: Probably a bigger TODO which is on the list of SCORPIO tasks is to create a generic case name for all runs that will be used for the output file names.  For example in EAM I think the pattern for all outputs is to have the name MACHINE.RESOLUTION.COMPSET.PROCESS.DATE.nc (or something along those lines)

File: components/scream/src/share/util/scream_family_tracking.hpp
Comment (jeff-cohere) [inline components/scream/src/share/util/scream_family_tracking.hpp:15]: Neat. Unless we want to introduce a "family" as a first-class concept, maybe we could call this `TypeTracking` or `TypeBasedTracking`?

File: components/scream/src/share/util/scream_family_tracking.hpp
Comment (jeff-cohere) [inline components/scream/src/share/util/scream_family_tracking.hpp:75]: Do or do not. There is no try (with the exception of exceptions).

File: components/scream/src/share/io/scream_output_manager.cpp
Comment (AaronDonahue) [inline components/scream/src/share/io/scream_output_manager.cpp:108]: Is there a way we can use the DAG or the actions during the DAG to add this label by default to all "required" fields?  I have a hacky way to do that in a separate WIP PR, see https://github.com/E3SM-Project/scream/pull/1155/files#diff-ffc785b3f4b9ce9afd2510be2ae5df1eaa5f92d3b18d0a7ca67f0b7e70aea5a2R132
But I'm sure this can be done better.

File: components/scream/src/share/io/scream_scorpio_interface.F90
Comment (AaronDonahue) [inline components/scream/src/share/io/scream_scorpio_interface.F90:1107]: The if and else are the same statements here.

File: components/scream/src/share/io/scream_scorpio_interface.F90
Comment (AaronDonahue) [inline components/scream/src/share/io/scream_scorpio_interface.F90:1160]: Same here, copy/paste

File: components/scream/src/control/atmosphere_driver.cpp
Comment (bartgol) [reply components/scream/src/control/atmosphere_driver.cpp:323]: I agree, I'll swap that around.

File: components/scream/src/control/atmosphere_driver.cpp
Comment (bartgol) [reply components/scream/src/control/atmosphere_driver.cpp:226]: Ah, yes, I change my mind when writing it, and left chunks of the first draft around. Will clean up.

File: components/scream/src/control/atmosphere_driver.hpp
Comment (bartgol) [reply components/scream/src/control/atmosphere_driver.hpp:89]: Good point.

File: components/scream/src/share/io/tests/io.cpp
Comment (AaronDonahue) [inline components/scream/src/share/io/tests/io.cpp:41]: This is commented out, should we delete, or is this a TODO for the next PR?

File: components/scream/src/share/field/field_header.hpp
Comment (jeff-cohere) [inline components/scream/src/share/field/field_header.hpp:105]: Does this comment apply to the two functions below?

File: components/scream/src/share/util/scream_family_tracking.hpp
Comment (bartgol) [reply components/scream/src/share/util/scream_family_tracking.hpp:4]: Argh, I saw that and wanted to get rid of it, then I forgot. You've got a 25/20 vision, sir.

File: components/scream/src/share/field/field_manager.hpp
Comment (jeff-cohere) [inline components/scream/src/share/field/field_manager.hpp:385]: bundled requests

File: components/scream/tests/uncoupled/p3/p3_standalone_output.yaml
Comment (AaronDonahue) [inline components/scream/tests/uncoupled/p3/p3_standalone_output.yaml:10]: We can add qv, qc, ... to this output file, now that you've fixed outputting tracers.

File: components/scream/src/share/io/scorpio_input.hpp
Comment (jeff-cohere) [inline components/scream/src/share/io/scorpio_input.hpp:42]: If people would rather have one field name per line (easier to insert/delete fields), one can use hyphens:

```
FIELDS:
    - field_name_1
    - field_name_2
    ...
    - field_name_N
```

File: components/scream/src/share/io/scorpio_input.hpp
Comment (jeff-cohere) [inline components/scream/src/share/io/scorpio_input.hpp:28]: You might stick a comma after `time` to make this read easier.

File: components/scream/src/share/io/scorpio_input.hpp
Comment (jeff-cohere) [inline components/scream/src/share/io/scorpio_input.hpp:77]: Do any of these constructors need explanation?

File: components/scream/src/share/util/scream_family_tracking.hpp
Comment (bartgol) [reply components/scream/src/share/util/scream_family_tracking.hpp:15]: I struggled with the name. I used 'family' only cause the class stores "parent" and "children", so I kept the analogy rolling. But a better name might help.

What about `PointerGraphTracking` or something with the word "graph" or "tree" in it?

Comment (bartgol) [reply components/scream/src/share/util/scream_family_tracking.hpp:15]: Technically, the graph might not be a tree, since I do not check that my parent is not me, nor one of my children (or grandchildren or...). In scream, the use case is quite limited, so I didn't feel the need for this check (though if used with circular deps, there would def be bugs). If we moved this file to ekat (which might be a possibility, given that it has no scream-based dep), then I might add checks.

File: components/scream/src/share/io/scream_output_manager.cpp
Comment (bartgol) [reply components/scream/src/share/io/scream_output_manager.cpp:108]: Yes, I wanted to get rid of this comment, and add code in the AD that deduces the restart group. I didn't get to that. If you want, I can do it here, otherwise we can do a separate PR (it's a very self-contained task, so I'd vote for PR, and assign to whomever got time).

File: components/scream/src/share/io/scream_scorpio_interface.F90
Comment (bartgol) [reply components/scream/src/share/io/scream_scorpio_interface.F90:1107]: Ctrl+C - Ctrl+V

File: components/scream/src/share/io/scream_output_manager.cpp
Comment (AaronDonahue) [reply components/scream/src/share/io/scream_output_manager.cpp:108]: We can keep it in PR 1155 since it's one of the first places it will be used explicitly and then when it's ready to go you can review if there is a better way then what I did.

File: components/scream/src/share/io/tests/io.cpp
Comment (bartgol) [reply components/scream/src/share/io/tests/io.cpp:41]: Oh, lol! I added that while debugging, and completely forgot. I am thankful I didn't follow my usual habit of filling debug messages/vars/names with "colorful" italian words... :D

File: components/scream/src/share/io/scream_scorpio_interface.F90
Comment (AaronDonahue) [reply components/scream/src/share/io/scream_scorpio_interface.F90:1107]: or for us VI users "yy" - "p"

File: components/scream/src/share/field/field_header.hpp
Comment (bartgol) [reply components/scream/src/share/field/field_header.hpp:105]: Ah, well, yes and no. You do need those two fcns to trigger `enable_shared_from_this` extra capabilities. But the second should be the only way to create a subfield header...except that I left the subfield header ctor declared (but never defined!).

I will revert back to singular, and make a separate comment for the 2nd fcn.

File: components/scream/src/share/io/tests/io.cpp
Comment (AaronDonahue) [reply components/scream/src/share/io/tests/io.cpp:41]: Your debug messages are helping me prepare for the next time I watch soccer with my in-laws...

File: components/scream/src/share/field/field_manager.hpp
Comment (bartgol) [reply components/scream/src/share/field/field_manager.hpp:385]: Ah, dang. I felt a bit like a "Jeff" for catching these typos, but I clearly can't outperform the master. Good catch!

File: components/scream/tests/uncoupled/p3/p3_standalone_output.yaml
Comment (bartgol) [reply components/scream/tests/uncoupled/p3/p3_standalone_output.yaml:10]: Yes, that might add a bit of testing to the IO infrastructure.

File: components/scream/src/share/io/scorpio_input.hpp
Comment (bartgol) [reply components/scream/src/share/io/scorpio_input.hpp:42]: Does that still map to a sequence, like the syntax I used in the comment? Nice!

File: components/scream/src/share/io/scorpio_input.hpp
Comment (bartgol) [reply components/scream/src/share/io/scorpio_input.hpp:77]: Probably. I will add some.

File: components/scream/src/share/io/scorpio_input.hpp
Comment (jeff-cohere) [reply components/scream/src/share/io/scorpio_input.hpp:42]: Yup. It's alternate syntax for a YAML sequence.

File: components/scream/src/share/util/scream_family_tracking.hpp
Comment (jeff-cohere) [reply components/scream/src/share/util/scream_family_tracking.hpp:15]: If it should be a tree to avoid circular refs, it seems like this check might be useful. And it may have a home in EKAT--the curiously recurring template pattern is not too bad a tax to pay for this kind of tracking.

File: components/scream/src/share/io/scream_output_manager.cpp
Comment (bartgol) [reply components/scream/src/share/io/scream_output_manager.cpp:94]: Let's add it to the TODO list then. It should be item 817191, right? =)

Jokes aside, a good v1.1 issue.

File: components/scream/src/share/io/scream_output_manager.hpp
Comment (jeff-cohere) [inline components/scream/src/share/io/scream_output_manager.hpp:72]: Consider `add_output` for a "verbier" method name.

File: components/scream/src/share/io/scorpio_output.cpp
Comment (jeff-cohere) [inline components/scream/src/share/io/scorpio_output.cpp:96]: Are all these parameter names written down somewhere?

File: components/scream/src/share/io/scorpio_output.cpp
Comment (jeff-cohere) [inline components/scream/src/share/io/scorpio_output.cpp:118]: `found` is always `true` in this context.

File: components/scream/src/share/io/scorpio_output.cpp
Comment (bartgol) [reply components/scream/src/share/io/scorpio_output.cpp:96]: I doubt it. I will add some docs in the output header.

File: components/scream/src/share/util/scream_family_tracking.hpp
Comment (bartgol) [reply components/scream/src/share/util/scream_family_tracking.hpp:15]: Ok. I might do this in a second time, cause I don't want to delay this PR too much, if that's ok. I will work on it right after it gets merged though.

File: components/scream/src/share/io/scorpio_output.cpp
Comment (bartgol) [reply components/scream/src/share/io/scorpio_output.cpp:118]: Good catch.

**Comment by bartgol** [comment]
@AaronDonahue @jeff-cohere I think I addressed most of your comments. I tried to add/improve the documentation, explaining the meaning of fcn args as well as input plist entries.

Would you mind taking another look?

File: components/scream/tests/uncoupled/homme/CMakeLists.txt
Comment (jeff-cohere) [inline components/scream/tests/uncoupled/homme/CMakeLists.txt:9]: Maybe okay to leave as "standalone". Standalone (or stand-alone) is an adjective, like "standalone HOMME", whereas to "stand alone" is a verb phrase.

Not a big deal, though.

**Review by jeff-cohere — APPROVED** [review]
Looks good to me!

File: components/scream/tests/uncoupled/homme/CMakeLists.txt
Comment (bartgol) [reply components/scream/tests/uncoupled/homme/CMakeLists.txt:9]: I was just trying to make it so that we use the same version. There was some "standalone" and some "stand_alone". I tossed a coin, and picked one.

File: components/scream/tests/uncoupled/p3/p3_stand_alone_output.yaml
Comment (AaronDonahue) [inline components/scream/tests/uncoupled/p3/p3_stand_alone_output.yaml:7]: nitpicky, and not necessary, but this might be a good place to use the `-` example for making long lists of fields.

**Review by AaronDonahue — APPROVED** [review]
Looks good!  Thank you @bartgol

File: components/scream/tests/uncoupled/p3/p3_stand_alone_output.yaml
Comment (bartgol) [reply components/scream/tests/uncoupled/p3/p3_stand_alone_output.yaml]: Done

File: components/scream/tests/uncoupled/homme/CMakeLists.txt
Comment (bartgol) [reply components/scream/tests/uncoupled/homme/CMakeLists.txt:9]: Done

**Comment by bartgol** [comment]
Adding WIP till I figure out why the test randomly hangs. Meanwhile, I found some bugs that got me wondering how is master passing. E.g., see bug fix in 1e0a082. The dofs offset we computed were completely wrong. How does pio not crash in master?

File: components/scream/tests/uncoupled/p3/p3_standalone_output.yaml
Comment (jeff-cohere) [inline components/scream/tests/uncoupled/p3/p3_standalone_output.yaml:7]: Looks great!

File: components/scream/src/share/io/scream_scorpio_interface.F90
Comment (AaronDonahue) [inline components/scream/src/share/io/scream_scorpio_interface.F90:322]: Should we remove this comment?

File: components/scream/src/share/io/scream_scorpio_interface.F90
Comment (AaronDonahue) [inline components/scream/src/share/io/scream_scorpio_interface.F90:329]: If DimCounter isn't used currently then I would say no, it isn't needed.

File: components/scream/src/share/io/scream_scorpio_interface.F90
Comment (AaronDonahue) [inline components/scream/src/share/io/scream_scorpio_interface.F90:405]: Again, should we delete this comment?

**Review by jeff-cohere — APPROVED** [review]
Looks great, @bartgol.

**Review by AaronDonahue — APPROVED** [review]

**Comment by AaronDonahue** [comment]
I approved the PR, looks good, and is a ton of work!  I had a few comments, but they shouldn't hold up merging this PR.

I did have a follow-up question, does this PR address the discrepancy we noticed regarding the DOF indices passed to PIO?  I'm guessing that will be the subject of a later PR?

**Review by AaronDonahue — APPROVED** [review]

**Pull request merged**
